astropy/wcs/utils.py

```diff
@@ -825,3 +826,237 @@ def pixel_to_pixel(wcs_in, wcs_out, *inputs):
                 outputs[ipix] = np.broadcast_to(pixel_outputs[ipix], original_shape)
 
     return outputs[0] if wcs_out.pixel_n_dim == 1 else outputs
+
+def _linear_wcs_fit(params, lon, lat, x, y, w_obj):
+    """Objective function for fitting linear terms.
+    Parameters
+    ----------
+    params : array
+        6 element array. First 4 elements are PC matrix, last 2 are CRPIX.
+    lon, lat: array
+        Sky coordinates.
+    x, y: array
+        Pixel coordinates
+    w_obj: `~astropy.wcs.WCS`
+        WCS object
+        """
+    cd = params[0:4]
+    crpix = params[4:6]
+
+    w_obj.wcs.cd = ((cd[0], cd[1]), (cd[2], cd[3]))
+    w_obj.wcs.crpix = crpix
+    lon2, lat2 = w_obj.wcs_pix2world(x, y, 1)
+    resids = np.concatenate((lon-lon2, lat-lat2))
+
+    return resids
+    
+
+def _sip_fit(params, lon, lat, u, v, w_obj, order, coeff_names):
+
+    """ Objective function for fitting SIP.
+     Parameters
+    -----------
+    params : array
+        6 element array. First 4 elements are PC matrix, last 2 are CRPIX.
+    lon, lat: array
+        Sky coordinates.
+    u, v: array
+        Pixel coordinates
+    w_obj: `~astropy.wcs.WCS`
+        WCS object
+    """
+
+    from ..modeling.models import SIP, InverseSIP   # here to avoid circular import
+
+    # unpack params
+    crpix = params[0:2]
+    cdx = params[2:6].reshape((2, 2))
+    a_params = params[6:6+len(coeff_names)]
+    b_params = params[6+len(coeff_names):]
```

Comment:
The docstring above suggests there are only 6 elements, but this suggests otherwise?

Reply:
correct, ill fix the docs. meant to say its at LEAST 6 elements, it depends on the degree of polynomial being fit.

Reply:
Ok sounds good!